src/github.com/getlantern/peerscanner/web.go

```diff
@@ -0,0 +1,150 @@
+
+	log.Debugf("About to listen at %v", laddr)
+	l, err := tls.Listen("tcp", laddr, tlsConfig)
+	if err != nil {
+		log.Fatalf("Unable to listen for tls connections at %s: %s", laddr, err)
+	}
+
+	log.Debug("About to serve")
+	err = http.Serve(l, nil)
+	if err != nil {
+		log.Fatalf("Unable to serve: %s", err)
+	}
+}
+
+// register is the entry point for peers registering themselves with the service.
+// If peers are successfully vetted, they'll be added to the DNS round robin.
+func register(resp http.ResponseWriter, req *http.Request) {
+	name, ip, port, err := getHostInfo(req)
+	if err == nil && port != 443 {
+		err = fmt.Errorf("Port %d not supported, only port 443 is supported", port)
+	}
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(resp, err.Error())
+		return
+	}
+
+	h := getOrCreateHost(name, ip)
+	online, connectionRefused, timedOut := h.status()
+	if timedOut {
+		log.Debugf("%v timed out waiting for status, returning 500 error", h)
+		resp.WriteHeader(500)
+		fmt.Fprintf(resp, "Timed out waiting for status")
+		return
+	}
+
+	if online {
+		resp.WriteHeader(200)
+		fmt.Fprintln(resp, "Connectivity to proxy confirmed")
+		return
+	}
+
+	// Note this may not work across platforms, but the intent
+	// is to tell the client if the connection was flat out
+	// refused as opposed to timed out in order to allow them
+	// to configure their router if possible.
+	if connectionRefused {
+		// 417 response code.
+		resp.WriteHeader(http.StatusExpectationFailed)
+		fmt.Fprintln(resp, "No connectivity to proxy - connection refused")
+	} else {
+		// 408 response code.
+		resp.WriteHeader(http.StatusRequestTimeout)
+		fmt.Fprintln(resp, "No connectivity to proxy - test request timed out")
+	}
+}
+
+// unregister is the HTTP endpoint for removing peers from DNS. Peers are
+// unregistered based on their ip (not their name).
+func unregister(resp http.ResponseWriter, req *http.Request) {
+	_, ip, _, err := getHostInfo(req)
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(resp, err.Error())
+		return
+	}
+
+	h := getHostByIp(ip)
+	msg := "Host not registered"
+	if h != nil {
+		h.unregister()
+		msg = "Host unregistered"
+	}
+	resp.WriteHeader(200)
+	fmt.Fprintln(resp, msg)
+}
+
+func getHostInfo(req *http.Request) (name string, ip string, port int, err error) {
+	name = req.FormValue("name")
+	if name == "" {
+		err = fmt.Errorf("Please specify a name")
+		return
+	}
+	ip = clientIpFor(req, name)
+	if ip == "" {
+		err = fmt.Errorf("Unable to determine IP address")
+		return
+	}
+	portString := req.FormValue("port")
+
+	if portString != "" {
+		port, err = strconv.Atoi(portString)
+		if err != nil {
+			err = fmt.Errorf("Received invalid port for %v - %v: %v", name, ip, portString)
+		}
+	}
+
+	return
+}
+
+func clientIpFor(req *http.Request, name string) string {
+	// Client requested their info
+	clientIp := req.Header.Get("X-Peerscanner-Forwarded-For")
```

Comment:
Third parties could send fake `X-Peerscanner-Forwarded-For` headers, and `clientIpFor()` will just trust the value the client is sending. Whatever this value is, it can be fed to `h := getOrCreateHost(name, ip)`. Should we worry about fake or badly formatted IPs, or about collisions with existent IPs (keys) on the `hostsByIp` map?